text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+- Start Date: 2016-11-05
+- RFC PR: https://github.com/emberjs/rfcs/pull/176
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Make Ember feel less overwhelming to new users, and make Ember applications
+start faster, by replacing the `Ember` global with a first-class system for
+importing just the parts of the framework you need.
+
+# Motivation
+
+ECMAScript 2015 (also known as ES2015 or ES6) introduced a syntax for importing
+and exporting values from modules. Ember aggressively adopted modules, and if
+you've used Ember before, you're probably familiar with this syntax:
+
+```js
+import Ember from "ember";
+import Analytics from "../mixins/analytics";
+
+export default Ember.Component.extend(Analytics, {
+  // ...
+});
+```
+
+One thing to notice is that the entire Ember framework is imported as a single
+package. Rather than importing `Component` directly, for example, you import
+`Ember` and subclass `Ember.Component`. (And this example still works even if
+you forget the import, because we also create a global variable called
+`Ember`.)
+
+Using Ember via a monolithic package or global object is not ideal for several
+reasons:
+
+* It's overwhelming for learners. There's a giant list of classes and functions
+  with no hints about how they're related. The API documentation reflects this.
+* Experienced developers who don't want all of Ember's features feel like
+  they're adding unnecessary and inescapable bloat to their application.
+* The `Ember` object must be built at boot time, requiring that we ship the
+  entire framework to the browser. This has two major costs:
+  1. Increased download time, particularly noticeable on slower connections.
+  2. Increased parsing/evaluation cost, which still must be paid even when
+     assets are cached. On some browsers/devices, this can far exceed the cost of the
+     download itself.
+* Running multiple versions of Ember on the same page is error-prone.
+
+Defining a public API for importing parts of Ember via JavaScript modules helps
+us lay the groundwork for solving all of these problems.
+
+#### Multiple Apps on a Page
```

Comment:
It's not even the "multiple apps" story that is the primary problem. It's the multiple inclusion of modules of _different_ versions of Ember which is a tremendously difficult problem to solve and requires things like symbol mangling. Otherwise all of the different versions of Ember will end up in the same namespace and get clobbered when we `overwrite: true` See: https://github.com/ef4/ember-browserify/pull/100/commits. This is, in effect, promising that we would use `rollup` (or something similar) to bundle individual versions of Ember (Ember1 ... EmberN) which would be wildly inefficient when we would instead want to only include particular versions of a smaller subset of the modules.

So, this is _entirely_ possible once we land this RFC, but it promises a lot that we're not sure we want to support. The goal would be to only have _one_ version of any browser dependency to constrain bytes sent over the wire.

I'd also be in favor of a rewrite to simply drop:

> its own version without the risk of clobbering another.

Because we in _no_ way guarantee that right now.